cirq-google/cirq_google/api/v2/ndarrays.py

```diff
@@ -0,0 +1,534 @@
+    """Convert a Float64Array proto to a numpy array of the same shape.
+
+    Args:
+        msg: Message to extract array from.
+
+    Returns:
+        Numpy array with 64-bit float type.
+    """
+    return _from_float_array(msg, "f8")
+
+
+def to_float32_array(
+    array: np.ndarray, out: ndarrays_pb2.Float32Array | None = None
+) -> ndarrays_pb2.Float32Array:
+    """Populate a new or provided Float32Array message with given numpy array.
+
+    Supports any float type taking <=4 bytes. If given e.g. a f2/16-bit float,
+    this will recast as a 32-bit float and serialize.
+
+    Args:
+        array: Array to add.
+        out: Message to add to. If None, create new message.
+    """
+    if out is None:
+        out = ndarrays_pb2.Float32Array()
+    endianness, dtype = _to_dtype(array.dtype, "f4")
+    out.shape[:] = array.shape
+    out.endianness = endianness
+    out.flat_bytes = array.astype(dtype, copy=False).tobytes()
+    return out
+
+
+def from_float32_array(msg: ndarrays_pb2.Float32Array) -> np.ndarray:
+    """Convert a Float32Array proto to a numpy array of the same shape.
+
+    Args:
+        msg: Message to extract array from.
+
+    Returns:
+        Numpy array with 32-bit float type.
+    """
+    return _from_float_array(msg, "f4")
+
+
+def to_float16_array(
+    array: np.ndarray, out: ndarrays_pb2.Float16Array | None = None
+) -> ndarrays_pb2.Float16Array:
+    """Populate a new or provided Float16Array message with given numpy array.
+
+    Args:
+        array: Array to add.
+        out: Message to add to. If None, create new message.
+    """
+    if out is None:
+        out = ndarrays_pb2.Float16Array()
+    endianness, dtype = _to_dtype(array.dtype, "f2")
+    out.shape[:] = array.shape
+    out.endianness = endianness
+    out.flat_bytes = array.astype(dtype, copy=False).tobytes()
+    return out
+
+
+def from_float16_array(msg: ndarrays_pb2.Float16Array) -> np.ndarray:
+    """Convert a Float16Array proto to a numpy array of the same shape.
+
+    Args:
+        msg: Message to extract array from.
+
+    Returns:
+        Numpy array with 16-bit float type.
+    """
+    return _from_float_array(msg, "f2")
+
+
+def _from_float_array(msg: TFloatMsg, dtype_base: npt.DTypeLike) -> np.ndarray:
+    if not msg.shape:
+        raise ValueError(f"Cannot convert unset/empty {type(msg)} message to a numpy array.")
+
+    array_format = msg.WhichOneof("data")
+    if array_format != "flat_bytes":
+        raise ValueError(
+            "Error deserializing ndarray proto: only flat_bytes format "
+            f"supported; given {array_format}"
+        )  # pragma: no cover
+
+    dtype = _from_dtype(cast(ndarrays_pb2.Endianness, msg.endianness), dtype_base)
+    flat = np.frombuffer(msg.flat_bytes, dtype=dtype)
+    return np.reshape(flat, msg.shape)
+
+
+V1IntMsgs = ndarrays_pb2.Int16Array | ndarrays_pb2.Int32Array | ndarrays_pb2.Int64Array
+V1UIntMsgs = ndarrays_pb2.UInt8Array
```

Comment:
We can follow the naming scheme for float and complex messages and put these definitions together at the top.
Also add the Int8Array

```suggestion
TIntMsg = TypeVar(
    "TIntMsg",
    ndarrays_pb2.Int8Array,
    ndarrays_pb2.Int16Array,
    ndarrays_pb2.Int32Array,
    ndarrays_pb2.Int64Array,
)
TUIntMsg = ndarrays_pb2.UInt8Array
```